76ers

twmoore3rd opened the pull request:
Update to PHP 7.6 using PDO for database access. Added to documentation. Functionality not changed beyond using PDO for database API (MySQL database). Contact me for IP address for volunteer registration page and phone number to call in on.

twmoore3rd committed:
Install Notes Update

twmoore3rd committed:
config.php runsw without outward complaint

#xxx??? used to mark fragments that need to be anonymized before fork merge

twmoore3rd committed:
index SQL updated to PDO

twmoore3rd committed:
Capture change made to confi.php

THIS FILE SHOULD NOT BE DISTRIBUTED

tmoorebetazi committed:
config.php and setup.php apparently pass start up tests; text sent to me by setup.php

tmoorebetazi committed:
ignore config.php and syntax correction

The configured config.php should not be uploaded due to security issues.

tmoorebetazi committed:
Undefined variables cleaning
Inserted 'VALUES' into SQL INSERT (D'oh)

twmoore3rd committed:
Swap in PDO for mysql calls

tmoorebetazi committed:
Delete index.original.php as not needed in code base

tmoorebetazi committed:
replaced mysql with PDO
corrected logAndDie error messages

tmoorebetazi committed:
Update .gitignore

tmoorebetazi committed:
Update .gitignore

tmoorebetazi committed:
Update .gitignore

tmoorebetazi committed:
Housekeeping
index.php:https commented out and left with a TODO

tmoorebetazi committed:
prepared statements syntax corrected
html cleaning
logic corrected for showing check boxes

tmoorebetazi committed:
Removed integer casts as they didn't do anything (error codes are already integers), removed "array('exception' => $e)" from the logAndDie calls as wasn't doing anything

tmoorebetazi committed:
Converted <? to <?php (long tags?); assured that all SQL statements were assigned to $sql and SQL keywords were CAPITALIZED; corrected to PDO exception

tmoorebetazi committed:
Replace <? short tag with <?php

tmoorebetazi committed:
called.php: inserted undefined variable (isset) protection
mainmenu.php: used long tags so php.ini doesn't need to be edited to support short tags
menu-digit.php: inserted undefined variable (isset) protection
sms.php: inserted undefined variable (isset) protection
verify.php: added echos with instructions on how to run script, fetchall used to get all rows

tmoorebetazi committed:
Fetch all the rows at once (memory is cheap :-) into an associated array. If there are retrieved records, dial the numbers

tmoorebetazi committed:
Noted required modifications with #TODO::to make it easier to find the necessary identifiers that need to be changed. The #TODO:: phrase is also nicely supported by PhpStorm

tmoorebetazi committed:
config.sample created; config.php is NOT pushed to github due to it containing passwords, etc.

twmoore3rd commented:
Error found in called.php therefore closing pull request until corrected.

tmoorebetazi committed:
Added comments and whitespace

twmoore3rd commented:
Re-opened PR with only changes to code being commenting and white spaces in called.php. called.php is apparently creating an error as it is trying to send texts to numbers that are not verified. Unfortunately number verification is not available with trial accounts. As the called.php code has not been modified, the assumption is being made that it works ergo the code is acceptable.

dustball commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
 .DS_Store
 *.bak
 *.txt
+/config.php
+/scratch.php
```
With a slash "/" in front, doesn't this reference files in the root of the drive, i.e. actual "/", where no files ever really are, just directories like  usr, bin, var, etc?

twmoore3rd replied:
Yep. My and/or PhpStorm's "mistake". 

Okay, let's call this sloppiness on my part. Will edit .gitignore to remove these two entries.

Oh good, .gitignore can have comments. :-)  I will comment that config.php should be ignored as one does not want to be uploading a fully configured, with passwords, etc. to GitHub. I know this first hand. How do I know first hand? I loaded an in-development config.php up to Github. No issues. And then 15 minutes later, Twilio emailed me saying that it had found my credentials in a repo and therefore had changed my authentication code.

The scratch.php was a scratch file trying out snippets. Both acceptable for a .gitignore file but should not be set up to ignore in the root directory.

When I told PhpStorm to ignore these two files, it somehow thought I was referring to files in the root.

dustball replied:
>  I will comment that config.php should be ignored as one does not want to be uploading a fully configured, with passwords, etc. to GitHub.

Yes.  Though `config.php` was already in the .gitignore file so that's odd.

dustball commented on .idea/dataSources.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Are the .idea files supposed to be checked in?  I don't know what they are or how they work.

Are they necessary for the program to run now?  How are they referenced from within the code?

twmoore3rd replied:
No, not necessary. Files left over from PhpStorm. With the exception of the .gitignore file, all the dot files and directories are not necessary for the code to run.

Will address.

dustball replied:
OK

twmoore3rd replied:
.gitignore now has a 
`./.idea/`
line that should ignore the .idea directory

dustball commented on README.md:
```diff
@@ -6,13 +6,13 @@
 This software creates a help line in the cloud which can be called by anybody that knows the number.  The current version is intended to be used by medium-to-large sized groups and organizations to support themselves, as opposed to a public help line which may have different needs.
 
-With just a web hosting account (that supports php 5.6 and MySQL) and paid Twilio account, you can launch a crisis line in an area code (and possibly number) of your choosing. Twilio costs $1 per month (for the phone number) plus 2 cents per minute (counting both legs of the connection). 
+With just a web hosting account (that supports php 5.6 and MySQL) and paid Twilio account (see Installation:Twilio setup notes below), you can launch a crisis line in an area code (and possibly number) of your choosing. Twilio costs $1 per month (for the phone number) plus 2 cents per minute (counting both legs of the connection). 
```
The supported PHP version(s) should be changed to ... ? do we know?

twmoore3rd replied:
7.x - good catch.

twmoore3rd replied:
In looking at the PR code, it *should* run in PHP 5.x and PHP 7.x in that the changes I made were around try/catch code and using PDO rather than MySQL calls. Both functionalities were available in PHP 5. As PHP 5.x installations have some degree of difficulty to install (have to track down code, not part of any available as part of any stock AWS AMIs though available in XAMPP/LAMPP downloads, etc.), how do you feel about the wording:

Been shown to work in PHP 7.x and is believed to run but has not been shown to run in PHP 5.6.

dustball commented on config.sample:
```diff
@@ -1,65 +1,107 @@
 
 # Help line config
 $anonymous = FALSE;                  # Setting to FALSE will inform both caller and volunteer who they talked to
-$help_line_number = "+16505551212";  # Must include prefix of +1 followed by phone number, no spaces or other characters
```
Why is all the sample data erased?  That is done to help the user quickly edit the file. 

When the new admin sees "+16505551212" they know exactly what format you they supposed to enter right away.  Forcing them to read the directions will at best slow them down, and at worst create a problem with a bad config.

twmoore3rd replied:
Differences in design philosophy. With assigned data, the "dear" user/implementer might accept these as defaults and the program would run though with unexpected results as +16505551212 is an appropriate formatted number but not the number that one might be expecting.

Tell ya what: go through and identify the sample data that I deleted and/or changed. I will then go through and test each of the variables against the "default" (properly formatted data though probably isn't the value the user wants). If the default is detected will echo an error for each of the defaults found. Once all the identified variables are tested and if an default value is found, the code will then exit. 

Good enough @dustball ? (just calling out to you on this one as it does need a reply from you)

Again, no ego on my part here and I am not calling you out in a negative way. Ah, the pleasures of working wit someone that you have no clue about. :-)

dustball replied:
Hmm. Let me think more about this.

One can programmatically check to see if the values are proper, and that will solve the "bad data" issue, but it doesn't address the "taking out sample data makes it harder to understand what to put in the first place and will increase the amount of time it takes to setup and potentially create a hurdle to usage" issue.  

As of right now I feel like removing sample data will result in it taking notably longer to fill in the file.  It looks intimidating and confusing without it IMHO.  But again lemme think about it.

twmoore3rd replied:
Uhm, sorry. Example - leave in:

`$help_line_number = "+16505551212";  # Must include prefix of +1 followed by phone number, no spaces or other characters`

Then in the setup.php, have the code check for $help_line_number being equal to "+16505551212". If equal, give fatal error message that $help_line_number in config.php must be set to an appropriate number.

Hey, wait, the config.php is sorta ugly now. What the lleh am I supposed to do says the implementer. What about comment having each assignment's comment start with #TODO:: and then details about the assignment statement. My #TODO comments starting in the first column would be deleted and your comments would stay in place but prefaced with the #TODO:: construct

dustball replied:
I'm going to repeat what you said in my own words to make sure I understand -

You propose:

- we go back to including the sample data
- we add extra checks to make sure the data is formatted correctly
- preface the existing comments with #TODO:: 

Is my understanding correct?  Apologies if not.  

All sounds good to me.

That all seems fine and extra checks surely can only make it better.  I think the most important one to check may be the db_charset since that is a new variable config variable and won't be populated by anyone migrating (if there are any existing installations of OCL in the wild, I'd probably know about it, but just in case ..).

dustball replied:
Also, the joke about the password showing up as stars only works with sample data of `hunter2`... 🤣

twmoore3rd replied:
A code fragment should say it all: :-)

<?php

$help_line_number = "+16505551212";  #TODO:: Must include prefix of +1 followed by phone number, no spaces or other characters
$friendly_phone = "(650) 555-HELP";  #TODO:: Human friendly version of the above number

$error_text = NULL;
if ("+16505551212" == $help_line_number)
    $error_text = $error_text."Don't mooch a non-exisitent number if you really want to help people\n\n";
if ("(650) 555-HELP" == $friendly_phone)
    $error_text = $error_text."$friendly_phone is not friendly enough\n\n";

if (NULL != $error_text){
    echo "Errors found in configuration data:\n\n$error_text\n";
    curlUpAndDie("FML or Sucks To Be You");
}
Hope some humor was successfully sent. 

Your original comments will taken verbatim and suffixed with "TODO:: " to make for easier searching, etc. Hmm, wonder if PHP has doc strings similar to Python. If so, I'll do a docstring with TODO embedded.

dustball commented on config.sample:
```diff
@@ -1,65 +1,107 @@
-    global $admin_email;
-    mail($admin_email,"[$system_name] Error","http://".$_SERVER["SERVER_NAME"].$_SERVER["REQUEST_URI"]."\n\n".print_r($_REQUEST,true)."\n\n\n".print_r($err,true));
+    global $system_name, $admin_email;
+#TODO::    mail($admin_email,"[$system_name] Error","http://".$_SERVER["SERVER_NAME"].$_SERVER["REQUEST_URI"]."\n\n".print_r($_REQUEST,true)."\n\n\n".print_r($err,true));
```
Should this be disabled by default?  Or is it better to have mail() call in there by default, and make a new config system to skip it if one needs?

twmoore3rd replied:
No, should not be disabled.

As you probably picked up, the #TODO:: is a flag used by PhpStorm so one can track work that needs to be done. And easy to grep on if not using PhpStorm. In this particular case, I meant to go back and uncomment it out. Didn't. Good catch.

dustball commented on index.php:
```diff
@@ -1,72 +1,121 @@
-    header('Location: ' . $location);
-    exit;
-}
+//TODO:: may want to keep commented out for development purposes, probably want to uncomment the code for production
```
The code should *always* be in production state.  If you need code for development, please wrap it with a config variable like $disable_https_redirect OR some other method (I think Apache exposes whether or not it has https enabled on the server etc).

dustball replied:
Just to add more color to my previous comment:

Right now, someone could install the code from the main branch and get a production quality product -- allbeit with the insane php 5.6 requirement.

I'd like to keep the "this is not beta software" vibe as much as possible.  

Exception: Once we finish with changes, we'll merge this in and it will in theory be production but will require testing to double check.  Thats an acceptable compromise IMHO.  

But we should't check in code that takes it from production-ready to software-in-development status, if that makes sense.

twmoore3rd replied:
Again, oversight on my part as I was trying to get this out (of my hair). Will uncomment the code but can't address if the code will run. As it ran in the past *and* it does not appear to be tied to a particular PHP version, should be alright.

dustball commented on index.php:
```diff
@@ -1,72 +1,121 @@
+    $txts = processREQUEST('txts');
+
+    try {  # make a prepared statement which also addresses escaping strings issues
+        $sql = "REPLACE INTO $table_name".              # delete record with phone primary key and then insert these values
```
That looks great, much better with the new API.

tmoorebetazi committed:
PR Corrections

ignore the .idea directory
activate the mail call
activate mandatory HTTPS. No certificate available will terminate index.php in an ugly way
Runs under PHP 7 on Linux and should/may run on other OSes and PHP 5

dustball commented on index.php:
```diff
@@ -111,62 +163,124 @@
   
-<? }  ?>
-
 <p><i><small>Powered by <a href="https://github.com/dustball/opencrisisline$volunteer">Open Crisis Line</a></small></i></p>
```
Oops, I introduced this bug.  Commenting here so I don't forget to fix. Or you can.

(Extraneous $volunteer, clearly the result of over-zealous search and replace...)

dustball commented on index.php:
```diff
@@ -1,72 +1,121 @@
-<title><? echo $system_name; ?></title>
+<title><?php echo $system_name; ?></title>
 <meta charset="utf-8">
 <link href="/jquery/jquery-ui.css" rel="stylesheet">
```
Another bug of mine.  Line calling for jquery should be removed.

dustball commented on .idea/.gitignore:
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
I see .idea/ added to .gitignore which is great, but I am still seeing these files in the PR?

twmoore3rd replied:
Good point.

Warning: a git n00b is in the wild. One of the issues with git that I have had for a while is that of what you speak: it automatically continues to track files that are no longer part of a release. :-( Good as time as ever to go back and see if I can get git/GitHub to totally forget about these files particularly as they were never part of a release.

Agree: with the exception of .gitignore, the 76er fork should not have any additional files than the master (convenient way of putting it).

twmoore3rd replied:
Believe both the .idea directory as well config.php have been deleted from the repo's complete history, i.e., there is no evidence of them the directory file from ever existing. 

Recorded here so I can come back to it some years from now: 
git filter-branch --force --index-filter \
  "git rm --cached --ignore-unmatch PATH-TO-YOUR-FILE-WITH-SENSITIVE-DATA" \
  --prune-empty --tag-name-filter cat -- --all
[GitHub](https://help.github.com/en/github/authenticating-to-github/removing-sensitive-data-from-a-repository) from the CLI.

`--prune-empty` and `--tag-name-filter` threw errors so deleted. As a directory was being deleted, use recursion:
`git filter-branch --force --index-filter "git rm --cached --ignore-unmatch .idea -r cat -- --all` to remove .idea directory and `git filter-branch --force --index-filter "git rm --cached --ignore-unmatch config.php -r cat -- --all` to remove config.php . 

Finally `git push origin --force --all` to push the changes back up on the repo.

dustball commented on README.md:
```diff
@@ -6,13 +6,13 @@
 This software creates a help line in the cloud which can be called by anybody that knows the number.  The current version is intended to be used by medium-to-large sized groups and organizations to support themselves, as opposed to a public help line which may have different needs.
 
-With just a web hosting account (that supports php 5.6 and MySQL) and paid Twilio account, you can launch a crisis line in an area code (and possibly number) of your choosing. Twilio costs $1 per month (for the phone number) plus 2 cents per minute (counting both legs of the connection). 
+With just a web hosting account supporting PHP (shown to work with PHP 7.x and  and is believed to run but has not been shown to run in PHP PHP 5.6) and MySQL and paid Twilio account (see Installation:Twilio setup notes below), you can launch a crisis line in an area code (and possibly number) of your choosing. Twilio costs $1 per month (for the phone number) plus 2 cents per minute (counting bot ...
```
> " believed to run but has not been shown to run in PHP PHP 5.6" 

?
